homeassistant/components/nextcloud/sensor.py

```diff
@@ -0,0 +1,172 @@
+    ],
+    "server_webserver": ["server", "webserver"],
+    "server_php_version": ["server", "php", "version"],
+    "server_php_memory_limit": ["server", "php", "memory_limit"],
+    "server_php_max_execution_time": ["server", "php", "max_execution_time"],
+    "server_php_max_file_size": ["server", "php", "upload_max_filesize"],
+    "server_database_type": ["server", "database", "type"],
+    "server_database_version": ["server", "database", "version"],
+    "server_database_size": ["server", "database", "size"],
+    "active_users_last_5_minutes": ["activeUsers", "last5minutes"],
+    "active_users_last_1_hour": ["activeUsers", "last1hour"],
+    "active_users_last_24_hours": ["activeUsers", "last24hours"],
+}
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None) -> None:
+    """Set up the Nextcloud sensor."""
+    # Fetch Nextcloud Monitor api data
+    ncm = NextcloudMonitor(
+        config[CONF_URL], config[CONF_USERNAME], config[CONF_PASSWORD]
+    )
+    if not ncm.data:
+        _LOGGER.error("No data returned from Nextcloud API. Check your user/password.")
+
+    hass.data[DOMAIN] = ncm.data
+
+    def nextcloud_update(event_time):
+        """Update data from nextcloud api."""
+        ncm.update()
+        hass.data[DOMAIN] = ncm.data
+
+    # Update sensors on time interval
+    track_time_interval(hass, nextcloud_update, config[CONF_SCAN_INTERVAL])
+
+    # Create list of sensors based on available nextcloud api data
+    sensors = []
+    for name, dict_path in NEXTCLOUD_ITEMS_EXPECTED.items():
+        sensors.append(NextcloudSensor(name, dict_path))
+
+    # Setup sensors
+    add_entities(sensors, True)
+
+
+class NextcloudSensor(Entity):
+    """Represents a Nextcloud sensor."""
+
+    def __init__(self, item, value_path):
+        """Initialize the Nextcloud sensor."""
+        self.item = item
+        self.value = None
+        self.value_path = value_path
+
+    @property
+    def icon(self):
+        """Return the icon for this sensor."""
+        if self.item[:6] == "system":
+            return "mdi:memory"
+        if self.item[:7] == "storage":
+            return "mdi:folder-account"
+        if self.item[:6] == "server":
+            return "mdi:server"
+        if self.item[:6] == "active":
+            return "mdi:account-multiple"
+        return "mdi:cloud"
+
+    @property
+    def name(self):
+        """Return the name for this sensor."""
+        return f"{DOMAIN}_{self.item}"
+
+    @property
+    def state(self):
+        """Return the state for this sensor."""
+        return self.value
+
+    @property
+    def unique_id(self):
+        """Return the unique ID for this sensor."""
+        return f"{DOMAIN}_{self.name}"
+
+    def update(self):
+        """Update the sensor."""
+        try:
+            data = self.hass.data[DOMAIN]
+            for path in self.value_path:
+                data = data[path]
+            self.value = data
+        except KeyError:
```

Comment:
Where and when can this `KeyError` occur? I am assuming it happens `data[path]` here since `self.hass.data[DOMAIN]` is always filled right?

Reply:
@springstan  Yes, `KeyError` could occur at `data[path]`, and yes, at this point `self.hass.data[DOMAIN]` should always be filled with something.

Reply:
IMHO you should only wrap `data[path]` in the try-catch-block.

Reply:
Ok, makes sense.